File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    
    let nickname = new Map();
    let messages = [];
    
    // 유저 아이디와 닉네임 매핑 관리
    for (let i = 0; i < record.length; i++) {
        const [action, uid, name] = record[i].split(" ");
        
        if (action === "Enter") {
            nickname.set(uid, name);
            messages.push({uid, action: "enter"});
        } else if (action === "Leave") {
            messages.push({uid, action: "leave"});
        } else if (action === "Change") {
            nickname.set(uid, name);
        }
    }
    
    // 출력 메시지 생성
    return messages.map(({ uid, action }) => {
        const name = nickname.get(uid);
        return `${name}님이 ${action === "enter" ? "들어왔습니다." : "나갔습니다."}`;
    });
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    username = {}
    
    for r in record:
        sr = r.split(" ")
        
        if sr[0] in ["Enter", "Change"]:
            username[sr[1]] = sr[2]
    
    for r in record:
        sr = r.split(" ")
        
        if sr[0] == "Enter":
            answer.append(username[sr[1]] + "님이 들어왔습니다.")
        elif sr[0] == "Leave":
            answer.append(username[sr[1]] + "님이 나갔습니다.")
            
    return answer